### PR TITLE
Support not array object

### PR DESCRIPTION
**Pull request by tksfjt1024**
Make it possible to do guess/preview/run even if JSON Path does not return array.

### CODE REVIEW

**Comment by hiroyuki-sato** [comment]
Thank you for crating this PR.
I'll check this later.
Could you write test code?
And, could you show me example input data?

**Comment by tksfjt1024** [comment]
I'll write test code later!

The following is a sample file and the result of `guess`.

### sample file (test.json)
```json
{
  "num": 1,
  "str": "hoge",
  "bool" : false
}
```

### seed.yml
```yml
in:
  type: file
  path_prefix: ./test.json
out:
  type: stdout

```

### guess result
```yml
in:
  type: file
  path_prefix: ./test.json
  parser:
    type: jsonpath
    charset: UTF-8
    newline: LF
    columns:
    - {name: num, type: long}
    - {name: str, type: string}
    - {name: bool, type: boolean}
out: {type: stdout}
```

**Comment by hiroyuki-sato** [comment]
I'll wait for adding a test. Please let me know if you want to merge quickly without a test at first.

BTW, Input Hash data always record just one entry, correct?
I can't imagine multiple records in this case.

**Comment by tksfjt1024** [comment]
I have no problem including the test in this PR. I'll write the test when I have time, so stay tuned!

>Input Hash data always record just one entry, correct?

You're right, I'm assuming the case where the whole Hash is considered as one record and transferred.

File: lib/embulk/guess/jsonpath.rb [hunk 1]
```diff
@@ -10,21 +10,18 @@ def guess_text(config, sample_text)
...
-            column[:format] = c.format if c.format
-            column
-          end
-          parser_guessed = {"type" => "jsonpath"}
-          parser_guessed["columns"] = columns
-          return {"parser" => parser_guessed}
-        else
-          raise RuntimeError,"Can't guess specified the JSONPath: #{json_path}. The results does not return an Array."
-        end
+        json = [json] unless json.is_a?(Array)
```
Comment (hiroyuki-sato) [inline lib/embulk/guess/jsonpath.rb:17]: Could you add comments like the below? (If you can write a more good description, please change it.)

```ruby
#The rest of code expect a data type like `[ {}, {}, {}]`.
#If JSONPath specifies a Hash object, it needs to pack a hash object into an array.
# ex. { "name": "bob", "age":24" } -> [ { "name": "bob", "age":24 } ]
```

File: src/main/java/org/embulk/parser/jsonpath/JsonpathParserPlugin.java [hunk 2]
```diff
@@ -172,37 +172,19 @@ public void run(TaskSource taskSource, Schema schema,
...
-                            visitor.setValue(value);
-                            column.visit(visitor);
+                        catch (DataException e) {
+                            skipOrThrow(e, stopOnInvalidRecord);
+                            continue;
                         }
-
-                        pageBuilder.addRecord();
+                    }
+                } else {
```
Comment (hiroyuki-sato) [inline src/main/java/org/embulk/parser/jsonpath/JsonpathParserPlugin.java]: I think this part needs to check whether `json` is an Object or not, like the below.
Does the current code handle properly the following input?

```json
[ { "name": "bob", "age":24 },{ "name":"alice", "age":25 } ]
```
* jsonpath: `$[1].name`  # bob

```java
} else ( json.isObject() ) { // please fix method name if this method incorrect.
    //
} else {
    // raise Exception.
}
```

**Review by hiroyuki-sato — CHANGES_REQUESTED** [review]
@tksfjt1024 Could you check my comments?

**Comment by chikamura** [comment]
@hiroyuki-sato 
Sorry for the very late reply. 🙇
I added tests and comments, so could you please check it?

File: src/main/java/org/embulk/parser/jsonpath/JsonpathParserPlugin.java (hunk 2 above)
Comment (chikamura) [reply src/main/java/org/embulk/parser/jsonpath/JsonpathParserPlugin.java]: Does `jsonpath: $[1].name` mean `root: $[1].name` ?
If json is neither an object nor an array, an error will occur here.
https://github.com/hiroyuki-sato/embulk-parser-jsonpath/blob/006de80b4035393f7502cc8462b79e26280f7984/src/main/java/org/embulk/parser/jsonpath/JsonpathParserPlugin.java#L215-L219

Please see also this test.
https://github.com/hiroyuki-sato/embulk-parser-jsonpath/blob/006de80b4035393f7502cc8462b79e26280f7984/src/test/java/org/embulk/parser/jsonpath/TestJsonpathParserPlugin.java#L475-L489

**Comment by hiroyuki-sato** [comment]
@tksfjt1024 @chikamura Thanks! I'll check this PR in a week. please wait.

File: src/test/java/org/embulk/parser/jsonpath/TestJsonpathParserPlugin.java [hunk 3]
```diff
@@ -439,6 +440,54 @@ public void writeNilsWithJsonPath()
...
+                "]"
+        ));
+
+        List<Object[]> records = Pages.toObjects(schema.toSchema(), output.pages);
+        assertEquals(1, records.size());
+        assertEquals("embulk", records.get(0)[0]);
+    }
+
+    @Test
+    public void rootPathJsonIsNotArrayOrObject() {
```
Comment (hiroyuki-sato) [inline src/test/java/org/embulk/parser/jsonpath/TestJsonpathParserPlugin.java]: Could you fix `src/test/java/org/embulk/parser/jsonpath/TestJsonpathParserPlugin.java:475:50: '{' at column 50 should be on a new line. [LeftCurly]`?

```suggestion
    public void rootPathJsonIsNotArrayOrObject()
    {
```

File: src/test/java/org/embulk/parser/jsonpath/TestJsonpathParserPlugin.java [hunk 4]
```diff
@@ -439,6 +440,54 @@ public void writeNilsWithJsonPath()
...
+        transaction(config, fileInput("{ \"_c0\": \"embulk\" }"));
+
+        List<Object[]> records = Pages.toObjects(schema.toSchema(), output.pages);
+        assertEquals(1, records.size());
+        assertEquals("embulk", records.get(0)[0]);
+    }
+
+    @Test
+    public void rootPathJsonIsObject()
+            throws Exception {
```
Comment (hiroyuki-sato) [inline src/test/java/org/embulk/parser/jsonpath/TestJsonpathParserPlugin.java]: Could you fix `/src/test/java/org/embulk/parser/jsonpath/TestJsonpathParserPlugin.java:459:30: '{' at column 30 should be on a new line. [LeftCurly]`?

```suggestion
    public void rootPathJsonIsObject()
            throws Exception
    {
```

File: src/main/java/org/embulk/parser/jsonpath/JsonpathParserPlugin.java (hunk 2 above)
Comment (hiroyuki-sato) [inline src/main/java/org/embulk/parser/jsonpath/JsonpathParserPlugin.java]: Could you fix the style error?

```
[ant:checkstyle] [ERROR] .../src/main/java/org/embulk/parser/jsonpath/JsonpathParserPlugin.java:185:17: '}' at column 17 should be alone on a line. [RightCurly]
```

```suggestion
                }
                else {
```

File: src/test/java/org/embulk/parser/jsonpath/TestJsonpathParserPlugin.java [hunk 5]
```diff
@@ -39,6 +39,7 @@
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.assertThrows;
```
Comment (hiroyuki-sato) [inline src/test/java/org/embulk/parser/jsonpath/TestJsonpathParserPlugin.java]: Could you fix import ordering? swap `assertThrows` place.

```
[ant:checkstyle] [ERROR] embulk-parser-jsonpath/src/test/java/org/embulk/parser/jsonpath/TestJsonpathParserPlugin.java:42: Wrong order for 'org.junit.Assert.assertThrows' import. [ImportOrder]
```

```suggestion
import static org.junit.Assert.assertThrows;
import static org.junit.Assert.assertTrue;
```

**Review by hiroyuki-sato — CHANGES_REQUESTED** [review]
Thank you for creating PR. Basically LGTM👍 Could you fix code styles?

**Comment by hiroyuki-sato** [comment]
@takumakanari Could you take a look at this PR when you get a chance?

**Review by hiroyuki-sato — APPROVED** [review]
@chikamura Thanks! LGTM 👍
Please wait for @takumakanari's review.

**Comment by chikamura** [comment]
@hiroyuki-sato 
Thank you for your review.
I fixed the code styles, so could you please review again.

**Comment by chikamura** [comment]
Sorry for our comments crossed.
I appreciate your approval.  🙏

**Review by takumakanari — APPROVED** [review]
LGTM 👍  (apologize for the late reply 🙏 )

**Comment by hiroyuki-sato** [comment]
@takumakanari Thanks!

**Comment by chikamura** [comment]
@takumakanari @tksfjt1024 
Many Thanks!

**Pull request merged**